src/crypto-providers/signingValidation.ts

```diff
@@ -0,0 +1,256 @@
+    switch (cert.type) {
+      case TxCertificateKeys.STAKING_KEY_REGISTRATION:
+        numStakeRegistrationItems += 1
+        break
+
+      case TxCertificateKeys.STAKING_KEY_DEREGISTRATION:
+      case TxCertificateKeys.DELEGATION:
+        numStakeOtherItems += 1
+        break
+
+      case TxCertificateKeys.STAKEPOOL_RETIREMENT:
+        numPoolColdItems += 1
+        break
+
+      default:
+        break
+    }
+  })
+  return { numStakeRegistrationItems, numStakeOtherItems, numPoolColdItems }
+}
+
+const validateOrdinaryTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  _validateStakeCredentialsAreAllOfType(
+    unsignedTxParsed, StakeCredentialType.ADDR_KEY_HASH, Errors.ScriptStakeCredentialInOrdinaryTx,
+  )
+}
+
+const validateOrdinaryWitnesses = (params: SigningParameters) => {
+  const {
+    stakeSigningFiles, poolColdSigningFiles, mintSigningFiles, multisigSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  const {
+    numStakeRegistrationItems, numStakeOtherItems, numPoolColdItems,
+  } = _countWitnessableItems(params.unsignedTxParsed)
+
+  if (numStakeRegistrationItems + numStakeOtherItems === 0 && stakeSigningFiles.length > 0) {
+    throw Error(Errors.TooManyStakeSigningFilesError)
+  }
+  if (numPoolColdItems === 0 && poolColdSigningFiles.length > 0) {
+    throw Error(Errors.TooManyPoolColdSigningFilesError)
+  }
+  if (!params.unsignedTxParsed.mint && mintSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMintSigningFilesError)
+  }
+  if (multisigSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMultisigSigningFilesError)
+  }
+}
+
+const validateOrdinarySigning = (params: SigningParameters) => {
+  validateOrdinaryWitnesses(params)
+  // these checks should be performed only with createSignedTx call (when the signing file set is
+  // expected to be complete) on top of validateOrdinaryWitnesses
+
+  const {
+    paymentSigningFiles, stakeSigningFiles, poolColdSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  if (paymentSigningFiles.length === 0) {
+    throw Error(Errors.MissingPaymentSigningFileError)
+  }
+
+  const {
+    numStakeOtherItems, numPoolColdItems,
+  } = _countWitnessableItems(params.unsignedTxParsed)
+
+  if (numStakeOtherItems > 0 && (stakeSigningFiles.length === 0)) {
+    throw Error(Errors.MissingStakeSigningFileError)
+  }
+  if (numPoolColdItems > 0 && (poolColdSigningFiles.length === 0)) {
+    throw Error(Errors.MissingPoolColdSigningFileError)
+  }
+}
+
+const validatePoolRegistrationTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  _validateStakeCredentialsAreAllOfType(
+    unsignedTxParsed, StakeCredentialType.ADDR_KEY_HASH, Errors.ScriptStakeCredentialInOrdinaryTx,
+  )
+
+  if (unsignedTxParsed.certificates.length !== 1) {
+    throw Error(Errors.MultipleCertificatesWithPoolRegError)
+  }
+  if (unsignedTxParsed.withdrawals.length > 0) {
+    throw Error(Errors.WithdrawalIncludedWithPoolRegError)
+  }
```

Comment:
Those are checked by hw wallets themselves.. Maybe it is of some value for Trezor (not sure if a protobuf message is to be sent to validate this, probably trezor connect does not check for this).

Reply:
I think Trezor Connect doesn't check this. Personally, I have no idea how frequent are situations that would be prevented by these checks (that is, of what value are the checks for the users, vs. how valuable it is for us to not duplicate them). But I imagine it doesn't really matter much.

As above, if we agree to remove these checks, this commit does that: https://github.com/vacuumlabs/cardano-hw-cli/pull/108/commits/d671eeed4eec8a43542bce01eb2fcd899206d1ec